File: app/components/forms/password_reset_popup.py
"""
Password Reset Popup Component

Contains the password reset popup component for the main page.
"""

from fasthtml.common import *


class PasswordResetPopup:
    """Password reset popup component."""
    
    def __init__(self, translations=None, versioned_url=None, language="nl"):
        """Initialize the password reset popup component."""
        self.translations = translations or {}
        self.versioned_url = versioned_url or (lambda x: x)
        self.language = language
    
    def get_text(self, key: str, default: str = "") -> str:
        """Get translated text for the given key."""
        return self.translations.get(key, default)
    
    def render(self):
        """Render the password reset popup."""
        return Div(
            # Modal backdrop
            Div(
                # Modal content
                Div(
                    # Modal header
                    Div(
                        H2(
                            self.get_text("set_new_password_title", "Set New Password"),
                            cls="text-2xl font-bold text-[#3D2E7C] mb-2"
                        ),
                        Button(
                            Img(
                                src=self.versioned_url("/static/assets/close.svg"),
                                alt="Close",
                                cls="w-6 h-6 filter brightness-0"
                            ),
                            id="close-password-reset-popup",
                            cls="text-gray-700 hover:text-gray-900 transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-[#3D2E7C] focus-visible:ring-offset-2 rounded-lg"
                        ),
                        cls="flex justify-between items-center mb-6"
                    ),
                    
                    # Modal body
                    Div(
                        # Error message container
                        Div(
                            id="reset-popup-error",
                            cls="hidden mb-4 p-3 bg-red-50 border border-red-200 rounded-lg text-red-700 text-sm w-full break-words"
                        ),
                        
                        # Success message container
                        Div(
                            id="reset-popup-success",
                            cls="hidden mb-4 p-3 bg-green-50 border border-green-200 rounded-lg text-green-700 text-sm w-full break-words"
                        ),
                        
                        # Password reset form
                        Form(
                            # Hidden token field
                            Input(
                                type="hidden",
                                id="reset-token",
                                name="token"
                            ),
                            
                            # Hidden username field for accessibility
                            Input(
                                type="text",
                                name="username",
                                autocomplete="username",
                                style="display: none;",
                                tabindex="-1",
                                aria_hidden="true"
                            ),
                            
                            Div(
                                Label(
                                    self.get_text("new_password_label", "New Password"), 
                                    for_="reset-new-password", 
                                    cls="block text-sm font-medium text-gray-700 mb-1"
                                ),
                                Input(
                                    type="password", 
                                    id="reset-new-password", 
                                    name="password", 
                                    required=True, 
                                    aria_required="true", 
                                    minlength="8",
                                    autocomplete="new-password",
                                    cls="w-full px-3 py-2 bg-[#F8F7FB] border border-gray-300 rounded-lg focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-[#3D2E7C] focus-visible:border-[#3D2E7C]"
                                ),
                                Div(
                                    self.get_text("password_min_length", "Password must be at least 8 characters long"), 
                                    cls="text-xs text-gray-500 mt-1"
                                ),
                                cls="flex flex-col gap-2 mb-4"
                            ),
                            
                            Div(
                                Label(
                                    self.get_text("confirm_password_label", "Confirm New Password"), 
                                    for_="reset-confirm-password", 
                                    cls="block text-sm font-medium text-gray-700 mb-1"
                                ),
                                Input(
                                    type="password", 
                                    id="reset-confirm-password", 
                                    name="confirm_password", 
                                    required=True, 
                                    aria_required="true",
                                    autocomplete="new-password",
                                    cls="w-full px-3 py-2 bg-[#F8F7FB] border border-gray-300 rounded-lg focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-[#3D2E7C] focus-visible:border-[#3D2E7C]"
                                ),
                                # Password mismatch error message
                                Div(
                                    self.get_text("passwords_mismatch", "Passwords do not match"),
                                    id="reset-password-mismatch-error",
                                    cls="hidden text-sm text-red-600 mt-1"
                                ),
                                cls="flex flex-col gap-2 mb-6"
                            ),
                            
                            Button(
                                self.get_text("reset_password_button", "Reset Password"),
                                type="submit",
                                id="reset-password-submit-btn",
                                data_loading_text=self.get_text("resetting", "Resetting..."),
                                data_default_text=self.get_text("reset_password_button", "Reset Password"),
                                cls="w-full bg-[#3D2E7C] hover:bg-[#3D2E7C]/90 text-white font-medium py-2 px-4 rounded-lg focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-[#3D2E7C] focus-visible:ring-offset-2 disabled:bg-gray-400 disabled:cursor-not-allowed disabled:hover:bg-gray-400"
                            ),
                            
                            id="reset-password-form",
                            method="post",
                            cls="flex flex-col",
                            role="form"
                        ),
                        cls=""
                    ),
                    
                    cls="bg-white rounded-lg shadow-xl p-6 w-full max-w-md mx-auto transform transition-all duration-300 ease-out scale-95 opacity-0"
                ),
                cls="fixed inset-0 bg-black/50 backdrop-blur-sm flex items-center justify-center p-4 z-50"
            ),
            id="password-reset-popup",
            cls="hidden"
        ) 

File: public/js/password-reset-popup.js
// Password reset popup functionality
document.addEventListener('DOMContentLoaded', function() {
    
    // Password reset popup elements
    const passwordResetPopup = document.getElementById('password-reset-popup');
    const passwordResetForm = document.getElementById('reset-password-form');
    const closePasswordResetBtn = document.getElementById('close-password-reset-popup');
    const submitButton = document.getElementById('reset-password-submit-btn');
    const passwordField = document.getElementById('reset-new-password');
    const confirmPasswordField = document.getElementById('reset-confirm-password');
    const errorMsg = document.getElementById('reset-password-mismatch-error');
    const tokenField = document.getElementById('reset-token');
    
    // Translation messages
    const messages = {
        nl: {
            passwordsNotMatch: 'Wachtwoorden komen niet overeen',
            networkError: 'Er is een fout opgetreden. Probeer het opnieuw.',
            resetting: 'Opnieuw instellen...',
            allFieldsRequired: 'Alle velden zijn verplicht',
            passwordTooShort: 'Wachtwoord moet minimaal 8 tekens bevatten'
        },
        en: {
            passwordsNotMatch: 'Passwords do not match',
            networkError: 'An error occurred. Please try again.',
            resetting: 'Resetting...',
            allFieldsRequired: 'All fields are required',
            passwordTooShort: 'Password must be at least 8 characters long'
        }
    };
    
    function getMessage(key) {
        const lang = TeambeeUtils.getCurrentLanguage();
        return messages[lang]?.[key] || messages.en[key] || key;
    }
    
    function openPasswordResetPopup(token) {
        if (!passwordResetPopup || !tokenField) return;
        
        // Set the token
        tokenField.value = token;
        
        // Close any other open popups/dropdowns
        TeambeeUtils.closeAllDropdowns();
        
        // Show popup
        TeambeeUtils.disableScroll();
        passwordResetPopup.classList.remove('hidden');
        
        setTimeout(() => {
            const modalContent = passwordResetPopup.querySelector('.bg-white');
            if (modalContent) {
                modalContent.style.transform = 'scale(1)';
                modalContent.style.opacity = '1';
            }
        }, 10);
        
        // Focus on password field
        setTimeout(() => {
            if (passwordField) passwordField.focus();
        }, 100);
    }
    
    function closePasswordResetPopup() {
        if (!passwordResetPopup) return;
        
        const modalContent = passwordResetPopup.querySelector('.bg-white');
        if (modalContent) {
            modalContent.style.transform = 'scale(0.95)';
            modalContent.style.opacity = '0';
        }
        
        setTimeout(() => {
            passwordResetPopup.classList.add('hidden');
            TeambeeUtils.enableScroll();
            
            // Clear form
            if (passwordResetForm) passwordResetForm.reset();
            if (tokenField) tokenField.value = '';
            
            // Hide messages
            TeambeeUtils.hideMessage(document.getElementById('reset-popup-error'));
            TeambeeUtils.hideMessage(document.getElementById('reset-popup-success'));
            
            // Reset validation
            if (errorMsg) errorMsg.classList.add('hidden');
            if (confirmPasswordField) confirmPasswordField.classList.remove('border-red-500');
        }, 200);
    }
    
    // Check if there's a reset_token in the URL
    const urlParams = new URLSearchParams(window.location.search);
    const resetToken = urlParams.get('reset_token');
    
    if (resetToken) {
        // Remove the token from URL for security
        const newUrl = window.location.pathname + (window.location.pathname.endsWith('/') ? '' : '/');
        window.history.replaceState({}, document.title, newUrl);
        
        // Open the password reset popup
        openPasswordResetPopup(resetToken);
    }
    
    // Close button handler
    if (closePasswordResetBtn) {
        closePasswordResetBtn.addEventListener('click', function(e) {
            e.preventDefault();
            closePasswordResetPopup();
        });
    }
    
    // Close popup when clicking outside
    if (passwordResetPopup) {
        passwordResetPopup.addEventListener('click', function(e) {
            if (e.target === passwordResetPopup || e.target.classList.contains('backdrop-blur-sm')) {
                closePasswordResetPopup();
            }
        });
    }
    
    // Close popup when pressing Escape key
    document.addEventListener('keydown', function(e) {
        if (e.key === 'Escape' && passwordResetPopup && !passwordResetPopup.classList.contains('hidden')) {
            closePasswordResetPopup();
        }
    });
    
    // Form validation
    function validatePasswordResetForm() {
        if (!passwordField || !confirmPasswordField || !submitButton) return false;
        
        const password = passwordField.value;
        const confirmPassword = confirmPasswordField.value;
        let isValid = true;
        
        // Check password length
        if (password.length < 8) {
            isValid = false;
        }
        
        // Check if passwords match
        if (password && confirmPassword && password !== confirmPassword) {
            confirmPasswordField.setCustomValidity(getMessage('passwordsNotMatch'));
            if (errorMsg) errorMsg.classList.remove('hidden');
            confirmPasswordField.classList.add('border-red-500');
            isValid = false;
        } else {
            confirmPasswordField.setCustomValidity('');
            if (errorMsg) errorMsg.classList.add('hidden');
            confirmPasswordField.classList.remove('border-red-500');
        }
        
        // Check if both fields are filled
        if (!password || !confirmPassword) {
            isValid = false;
        }
        
        // Update submit button state
        if (isValid) {
            submitButton.disabled = false;
            submitButton.classList.remove('disabled:bg-gray-400', 'disabled:cursor-not-allowed');
            submitButton.classList.add('bg-[#3D2E7C]', 'hover:bg-[#3D2E7C]/90');
        } else {
            submitButton.disabled = true;
            submitButton.classList.add('disabled:bg-gray-400', 'disabled:cursor-not-allowed');
            submitButton.classList.remove('hover:bg-[#3D2E7C]/90');
        }
        
        return isValid;
    }
    
    // Add validation listeners
    if (passwordField && confirmPasswordField) {
        passwordField.addEventListener('input', validatePasswordResetForm);
        confirmPasswordField.addEventListener('input', validatePasswordResetForm);
        
        // Initial validation
        validatePasswordResetForm();
    }
    
    // Form submission handler
    if (passwordResetForm) {
        passwordResetForm.addEventListener('submit', async function(e) {
            e.preventDefault();
            
            if (!validatePasswordResetForm()) {
                TeambeeUtils.showMessage(document.getElementById('reset-popup-error'), getMessage('allFieldsRequired'));
                return;
            }
            
            const token = tokenField ? tokenField.value : '';
            if (!token) {
                TeambeeUtils.showMessage(document.getElementById('reset-popup-error'), 'Invalid reset token');
                return;
            }
            
            // Show loading state
            submitButton.disabled = true;
            submitButton.textContent = submitButton.dataset.loadingText || getMessage('resetting');
            TeambeeUtils.hideMessage(document.getElementById('reset-popup-error'));
            TeambeeUtils.hideMessage(document.getElementById('reset-popup-success'));
            
            try {
                const formData = new FormData(passwordResetForm);
                formData.append('token', token);
                
                // Detect current page language and use appropriate endpoint
                const currentLang = TeambeeUtils.getCurrentLanguage();
                const resetUrl = currentLang === 'en' ? `/en/reset-password/${token}` : `/reset-password/${token}`;
                
                const response = await fetch(resetUrl, {
                    method: 'POST',
                    body: formData,
                    headers: {
                        'X-Requested-With': 'XMLHttpRequest'
                    }
                });
                
                const result = await response.json();
                
                if (result.success) {
                    TeambeeUtils.showMessage(document.getElementById('reset-popup-success'), result.message);
                    
                    // Clear form
                    passwordResetForm.reset();
                    
                    // Close popup after a short delay
                    setTimeout(() => {
                        closePasswordResetPopup();
                        
                        // Show success message on main page if available
                        const mainSuccessContainer = document.querySelector('.success-notification');
                        if (mainSuccessContainer && result.message) {
                            // You can implement a main page notification here if needed
                        }
                    }, 2000);
                } else {
                    TeambeeUtils.showMessage(document.getElementById('reset-popup-error'), result.message);
                    
                    // Clear password fields and focus on first field
                    if (passwordField) passwordField.value = '';
                    if (confirmPasswordField) confirmPasswordField.value = '';
                    if (passwordField) passwordField.focus();
                }
            } catch (error) {
                console.error('Error:', error);
                TeambeeUtils.showMessage(document.getElementById('reset-popup-error'), getMessage('networkError'));
            } finally {
                submitButton.textContent = submitButton.dataset.defaultText || 'Reset Password';
                validatePasswordResetForm(); // Re-validate to restore button state
            }
        });
    }
}); 